### DISCUSSION TITLE
STM32H7xx Reference Map questions

### DISCUSSION THREAD

**Discussion by phil-barrett** [grblHAL/core#404] [Q&A]
I've been playing with the STM32H7xx (H723 on a nucleo specifically) and have a few questions about the reference map (from grblHAL Web Builder). I'm very much a neophyte with STM so please forgive amateur questions.

1. I noticed that Aux Inputs are commented out.  Is there a specific reason for this? I noticed that inputs 0-2 use the same pins as output 4-6.
2. Safety door is not enabled if more than 3 motors. I believe that this is because M3 motor limit pin uses GPIOB:10. Is there a reason that safety door needs to be GPIOB:10? Wouldn't it make more sense to have the door on GPIOF:9 (M7 limit pin)? Then the door test can be changed to only disallow it for an 8 motor configuration.
3. SDCARD enable is commented out - Is SDMMC preferred on the H7xx? Is that supported by grblHAL?
4. I'm a little unclear on the differences between the Nucleo H7 boards (743ZI vs H723ZG) as it pertains to the reference map which is specifically for the H743ZI.  The ST docs are confusing to me. Any insights are greatly appreciated.

Thanks, 
Phil

**Comment by dresco** [comment]
Hi Phil,

1& 2 I've tried to keep it in line with Terje's F7 map, so is just reflecting the current state of the F7 driver..

3 Yes, have only implemented SDMMC in the H7 driver, as all the boards I have are using 4-bit wide SDMMC access.

4 Good spot. The reference map is used for both those Nucleo boards, I will update the description on next commit.

In terms of differences, the 723 has less flash and less RAM, and the potential to run a bit faster (550MHz instead of 480MHz). Currently, both builds are running at 480 for consistency.

I believe it is a pin compatible replacement, as has been substituted in most of the import development & 3D printer boards during the chip shortages. There are a few differences under the covers, but probably the most obvious is that the 723 does not have a USB FS peripheral, so uses the HS peripheral (running in FS mode). That was the only significant code change required to add support.

Cheers,
Jon.

**Reply by phil-barrett** [reply]
Thanks for the quick reply.

**Reply by phil-barrett** [reply]
I've been looking at the difference between the F723 and H723 map and there is one difference that I don't know if intentional.  I was going to post a diff list but it's kind of meaningless because all you would see are /* and */.  Basically the aux inputs are commented out in place of aux outputs 4, 5 and 6 which use the same pins.

**Reply by dresco** [reply]
No not intentional, the F7 driver was updated a few days ago, and I've just not merged those changes in yet...

**Reply by terjeio** [reply]
The F7 reference map needs work - my plan is/was to desgn a board for the Nucleo-144 form factor that uses all the pins: 8 axes, plenty of I/O, at least 3 UARTs, CanBus and what not. I have a sheet of paper somehere in my stack outlining it...

**Reply by phil-barrett** [reply]
The H723 appears to be quite a bit cheaper than the F723, both in LQFP-144. ($4 vs   I haven't crawled through the datasheets side by side but they are pretty similar.  JLC has this pricing:
> [STM32H723ZGT6]   1 + $4.8720 
> [STM32F723ZET6]   1 + $9.2250

Oddly enough the cheaper H part has 1024KB flash and the more expensive F part only has 512KB.  Not sure what in the F part justifies the higher price.

**Reply by terjeio** [reply]
> Not sure what in the F part justifies the higher price.

Availability?

Note that the [H723](https://www.st.com/en/microcontrollers-microprocessors/stm32h723zg.html) also has a double precision FPU unlike the F723. In the future I want to support that.

**Reply by phil-barrett** [reply]
JLC has 9000+ of the H and only 1000 of the F so that might be the reason.  LCSC shows similar stock levels and pricing.

**Reply by dresco** [reply]
That H723 price is pretty nuts, around $10 off the low quantity ST price!

And it's actually cheaper than the 100 pin equivalent, I've been struggling a bit with pins while working on my H7 board schematic - am now thinking I should just pivot to the larger IC and save some headaches..

**Reply by phil-barrett** [reply]
Checked mouser - the STM32H723ZGT6 is $13.12 Q1 and $8.31 Q1080 (yeah, weird quantity). Not sure how the Chinese got such a huge discount. Maybe it's down to good negotiation skills.  I doubt the STM32H723 is counterfeited, yet anyway.

**Reply by hanke-cnc** [reply]
The STM32H723ZGT6 is priced at only $3.6 in China. I am not sure why there is such a significant price difference between domestic and international markets. A typical example is the STM32F402RCT6, which is not available on the official ST website but can be found for only $1, with a considerable amount of stock.

**Reply by phil-barrett** [reply]
Where are you finding it for that price?  Be aware that there is a lot of counterfeiting going on.  Especially for the F1xx and F4xx series.  There are at least 7 F103 variants, some of which are close (but not perfect) and some of which are quite a bit different. And plenty of unscrupulous suppliers willing to remark a chip to get a better price. They may run gbrlHAL, or they may not.  Note also that a lot of the Chinese suppliers on Ali-whatever substitute similar product without ever telling you.  So if you buy something that works once, there is no guarantee you will the same thing the next time. Authorized distributors like LCSC seem to be fairly honest, by the way.

**Reply by hanke-cnc** [reply]
I am Chinese, and I purchase from Taobao <https://www.taobao.com/>. Perhaps you cannot buy from here? I am well aware of the situation here. Most Taobao stores sell chips that have been disassembled and refurbished, such as the STM32F407VET6 for $1. I am using it to produce a GRBLHAL control board. Of course, it comes with no after-sales guarantee. Besides, there are also some stores that only sell original authentic chips, and they are reliable. Currently, a new original STM32F407VET6 is around $2, and the STM32H750VBT6 has also reached $2. In addition, due to the chip shortage caused by the pandemic, many MCU companies based on ARM cores have emerged in China. They often have higher clock frequencies, lower prices, larger flash and ARM capacities than competing chips from ST. However, their software and documentation are often quite poor. Unless your usage is exceptionally large, there is no need to consider them.

**Reply by hanke-cnc** [reply]
Translate to English: Please convert the prices in the picture from Chinese Yuan to your familiar US dollars. Some popular chips are very affordable, while niche chips are still quite expensive.
![A](https://github.com/grblHAL/core/assets/108179425/a359613a-0eca-4844-a7e6-5b7a42414390)
![B](https://github.com/grblHAL/core/assets/108179425/b7acd3f6-c02b-47c7-b7e7-f7e7b640ffdb)
![C](https://github.com/grblHAL/core/assets/108179425/1de1f440-d944-4ffe-906b-6d7d2d32e046)

**Reply by phil-barrett** [reply]
Thanks for the reply.  Those are good prices.   taobao is hard for a non-Chinese speaker to use.

Yes, there are a lot of STM32 clones and many more coming. GD has a number of close clones.
